### PR TITLE
fixing duplicate connections after cleanup

### PR DESCRIPTION
**Pull request by abdelr**
### What's being changed:

We add duplicates after the cleanup tombstone logic because we simply add back connections without checking if they already existed. This PR fixes this issue.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [X] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/vector/hnsw/neighbor_connections.go [hunk 1]
```diff
@@ -321,11 +321,17 @@ func (n *neighborFinderConnector) connectNeighborAtLevel(neighborID uint64,
 
 	maximumConnections := n.maximumConnections(level)
 	if len(currentConnections) < maximumConnections {
-		// we can simply append
-		// updatedConnections = append(currentConnections, n.node.id)
-		neighbor.appendConnectionAtLevelNoLock(level, n.node.id, maximumConnections)
-		if err := n.graph.commitLog.AddLinkAtLevel(neighbor.id, level, n.node.id); err != nil {
-			return err
+		alreadyConnected := false
```
Comment (jeroiraz) [inline adapters/repos/db/vector/hnsw/neighbor_connections.go:324]: couldn't this validation (alreadyConnected) be done even when len(currentConnections) < maximumConnections doesn't hold?

Comment (abdelr) [reply adapters/repos/db/vector/hnsw/neighbor_connections.go]: It actually should. More than that, I added a second test to actively check if we have duplicates (even though the maxConnections is not violated) and we still had some. The second branch was adding just in rare occasions but still. Also, by adding this test, it is not flaky failing anymore. It consistently fails before the fix now.

File: adapters/repos/db/vector/hnsw/index_too_many_links_bug_integration_test.go [hunk 2]
```diff
@@ -9,9 +9,6 @@
 //  CONTACT: hello@weaviate.io
 //
 
-//go:build integrationTest && !race
```
Comment (jeroiraz) [inline adapters/repos/db/vector/hnsw/index_too_many_links_bug_integration_test.go:12]: if it's intentional the file might be renamed for consistency

File: adapters/repos/db/vector/hnsw/index_too_many_links_bug_integration_test.go [hunk 3]
```diff
@@ -263,6 +260,21 @@ func Test_NoRace_ManySmallCommitlogs(t *testing.T) {
 		}
 	})
 
+	t.Run("verify that the loaded index has no duplicated connections", func(t *testing.T) {
```
Comment (andrewisplinghoff) [inline adapters/repos/db/vector/hnsw/index_too_many_links_bug_integration_test.go:263]: Shouldn't we wait for the tombstone cleanup to finish before running this part of the test? When testing here without your fix there were no duplicates yet because the tombstone cleanup was still running.
